### PR TITLE
raster_temporal_mean_agg

### PR DESCRIPTION
**Pull request by jdroenner**

### CODE REVIEW

**Comment by jdroenner** [comment]
[Mean über Raster-Kacheln](https://trello.com/c/0jynZOHJ/206-mean-%C3%BCber-raster-kacheln)

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+impl<T> TemporalMeanTileAccu<T> {
+    pub fn add_tile(&mut self, in_tile: RasterTile2D<T>) -> Result<()>
+    where
+        T: Copy + AsPrimitive<f64> + Pixel,
+    {
+        self.time = self.time.union(&in_tile.time)?;
+
+        let in_tile_grid = match in_tile.grid_array {
+            GridOrEmpty::Grid(g) => g,
+            GridOrEmpty::Empty(_) => return Ok(()),
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:56]: Isn't Empty like adding NODATA?

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        };
+
+        match &mut self.value_grid {
+            GridOrEmpty::Empty(_) => {
+                let mut accu_grid = self.value_grid.clone().into_materialized_grid();
+
+                for ((av, ac), iv) in accu_grid
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:63]:
```suggestion
                for ((acc_value, acc_count), new_value) in accu_grid
```

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                        *av = ivf;
+                        *ac = 1;
+                    }
+                }
+
+                self.value_grid = accu_grid.into();
+            }
+
+            GridOrEmpty::Grid(accu_grid) => {
+                for ((av, ac), iv) in accu_grid
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:82]:
```suggestion
                for ((acc_value, acc_count), new_value) in accu_grid
```

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+impl<T> FoldTileAccu for TemporalMeanTileAccu<T>
+where
+    T: Pixel,
+{
+    type RasterType = T;
+
+    fn tile_ref(&self) -> &RasterTile2D<Self::RasterType> {
+        unimplemented!();
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:118]: Dangerous, could crash our app.
Can't you throw some error or output something meanwhile and add a TODO?

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        } = self;
+
+        let value_grid = match value_grid {
+            GridOrEmpty::Grid(g) => g,
+            GridOrEmpty::Empty(_) => {
+                return RasterTile2D::new(
+                    time,
+                    tile_position,
+                    global_geo_transform,
+                    EmptyGrid2D::new(value_grid.grid_shape(), out_no_data_value).into(),
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:139]: Can you just use the values within `GridOrEmpty::Empty(_)`?

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+        let res: Vec<T> = value_grid
+            .data
+            .into_iter()
+            .zip(count_grid.data.into_iter())
+            .map(|(v, c)| {
+                if c == 0 {
+                    out_no_data_value
+                } else {
+                    T::from_(v) // lets hope that this works..
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:152]: Why shouldn't it work? You created tests for it.

Or do you mean we have f64 and now its u8?

File: operators/src/processing/temporal_raster_aggregation/min_max_first_last_subquery.rs [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+                GridOrEmpty::Grid(a)
+            }
+            (GridOrEmpty::Empty(e), _) | (_, GridOrEmpty::Empty(e)) => GridOrEmpty::Empty(e),
+        }
+    };
+
+    accu_tile.grid_array = grid;
+    TemporalRasterAggregationTileAccu {
+        accu_tile,
+        initial_state: false,
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/temporal_raster_aggregation/min_max_first_last_subquery.rs:153]: Can't you use an empty grid as initial state?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: datatypes/src/raster/raster_tile.rs [hunk 8]
```diff
@@ -182,6 +182,15 @@ where
             properties: RasterProperties::default(),
         }
     }
+
+    pub fn materialize(&mut self) {
+        match self.grid_array {
+            GridOrEmpty::Grid(_) => {}
+            GridOrEmpty::Empty(_) => {
+                self.grid_array = self.grid_array.clone().into_materialized_grid().into();
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/raster_tile.rs:190]: we only clone if it is empty and use the existing transformation logic to generate a new Grid.

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs (hunk 4 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:118]: yes i was going to refactor the trait :(

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs (hunk 6 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:152]: yes i was thinking about f64 to u8

File: operators/src/processing/temporal_raster_aggregation/min_max_first_last_subquery.rs (hunk 7 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/min_max_first_last_subquery.rs:153]: nope then you cant know if you already had an empty tile

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs (hunk 1 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:56]: depends if it is the initial case

Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:56]: we need the         initial_state: false, logic i guess

Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs]: 5708885

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs (hunk 2 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs]: 5708885

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs (hunk 4 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs]: 5708885

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs (hunk 5 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs:147]: nope the accu tile is f64 and the result tile is T

File: operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs (hunk 3 above)
Comment (jdroenner) [reply operators/src/processing/temporal_raster_aggregation/mean_aggregation_subquery.rs]: 5708885

**Review by ChristianBeilschmidt — APPROVED** [review]

**Pull request merged**